hudi-spark/src/main/java/org/apache/hudi/keygen/CustomKeyGenerator.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.keygen;
+
+import org.apache.hudi.DataSourceWriteOptions;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.config.TypedProperties;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hudi.exception.HoodieDeltaStreamerException;
+import org.apache.hudi.exception.HoodieKeyException;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.noPartition;
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.simple;
+import static org.apache.hudi.keygen.KeyGenerator.PartitionKeyType.timestampBased;
+
+/**
+ * This is a generic implementation of KeyGenerator where users can configure record key as a single field or a combination of fields.
+ * Similarly partition path can be configured to have multiple fields or only one field. This class expects value for prop
+ * "hoodie.datasource.write.partitionpath.field" in a specific format. For example:
+ *
+ * properties.put("hoodie.datasource.write.partitionpath.field", "field1:PartitionKeyType1,field2:PartitionKeyType2").
+ *
+ * The complete partition path is created as <value for field1 basis PartitionKeyType1>/<value for field2 basis PartitionKeyType2> and so on.
+ *
+ * Few points to consider:
+ * 1. If you want to customise some partition path field on a timestamp basis, you can use field1:timestampBased
```

Comment:
all this needs to be documented for the user?

Reply:
in our documentation you mean to say? Happy to do that @vinothchandar 
Will raise a follow up PR once this gets merged.

Reply:
yes.. 

is the CustomKeyGenerator compatible with the `SimpleKeyGenerator` configs? I am wondering if we can replace the default with this, without forcing user to do any additional work.. I think this is worth pursuing.. (We can then rename this `DefaultKeyGenerator`

Reply:
No the configs are not compatible, since CustomKeyGenerator expects partitionPathFields to be provided in a particular format. :( 
But since we are going with a major release next, I guess we can make this as the default? (My thinking here is users can expect a bit of breaking changes in major releases, anyways we will mention all the changes in the release notes). WDYT?

Reply:
Also because this key generator pretty much covers all the possible cases for key generation.